spec/avram/time_criteria_spec.cr

```diff
@@ -14,6 +14,11 @@ describe Time::Lucky::Criteria do
     end
   end
 
+  it "as_date" do
+    input_date = "2012-01-31"
+    activated_at.as_date.eq(input_date).to_sql.should eq ["SELECT users.id, users.created_at, users.updated_at, users.activated_at FROM users WHERE DATE(users.activated_at) = $1", input_date]
```

Comment:
If someone passes a Time object to input date, will it do the right thing?

Reply:
It actually depends on what is the right thing to do. At this moment, it will not be compiling, with this error

`Error: expected argument #1 to 'String::Lucky.parse' to be Array(String), Nil or String, not Time`

The problem is that the function silently switches the criteria to a string. It's mostly about making a choice on how to represent its outcome in Lucky. Date does not exists and unfortunately, that's what the function is returning. So I had to choose between
- String
- Time
Now the problem becomes that the `eq()` expects the other to be of the same type (thus the error), so I had to choose between `.as_date.gte("2011-01-01")` or `as_date.gte(Time.utc)`, I opted for the first but I think it's really a matter of choices in there

Reply:
My initial instinct is that it would run the proper .to_s on the Time to get the date out as a string. I can conceive of ways that might not actually be the right thing to do -- though it is what I would have expected.

If there's enough ambiguity in the use case then sure, raising an error is probably the right thing to do. It's a good candidate for a more helpful error message rendered at compile time. If it's possible to lend advice on what actually is expected, that would be a good way to lead folks to the happy path. Eg "Did you mean .eq(time.psql_format)?" or whatever method it would actually be.

Reply:
There was an other alternative I was thinking about, but it risks becoming quite complicated after a while...
It basically consists of defining a new `Date` class, almost invisible to everyone but the Time::Lucky one, which would then redefine basically all operators with overloading (something like instead of having:
`criteria(T).eq(other : T)` we would have `criteria(Date).eq(other : String)` and `criteria(Date).eq(other : Time)`.
But honestly I don't even know if this is feasible

Reply:
Neat idea, but it can wait. Your scope here is nicely consumable as is.